src/raftstore/store/worker/read.rs

```diff
@@ -0,0 +1,902 @@
+                "{} leader lease is expired: {:?}",
+                self.delegate.tag, self.delegate.leader_lease
+            );
+        } else {
+            debug!("{} leader lease is None", self.delegate.tag,);
+            self.metrics.rejected_by_no_lease += 1;
+        }
+        LeaseState::Expired
+    }
+}
+
+impl<C: Sender<StoreMsg>> Runnable<Task> for LocalReader<C> {
+    fn run(&mut self, _: Task) {
+        unreachable!()
+    }
+
+    fn run_batch(&mut self, tasks: &mut Vec<Task>) {
+        let mut sent = None;
+        self.metrics.batch_requests_size.observe(tasks.len() as _);
+
+        for task in tasks.drain(..) {
+            match task {
+                Task::Register(delegate) => {
+                    debug!(
+                        "{} register ReadDelegate for {:?}",
+                        delegate.tag, delegate.peer_id
+                    );
+                    self.delegates.insert(delegate.region.get_id(), delegate);
+                }
+                Task::Read(StoreMsg::RaftCmd {
+                    send_time,
+                    request,
+                    callback,
+                }) => {
+                    self.propose_raft_command(request, callback, send_time);
+                    if sent.is_none() {
+                        sent = Some(send_time);
+                    }
+                }
+                Task::Read(StoreMsg::BatchRaftSnapCmds {
+                    send_time,
+                    batch,
+                    on_finished,
+                }) => {
+                    self.propose_batch_raft_snapshot_command(batch, on_finished);
+                    if sent.is_none() {
+                        sent = Some(send_time);
+                    }
+                }
+                Task::Read(other) => {
+                    unimplemented!("unsupported Msg {:?}", other);
+                }
+                Task::Update((region_id, progress)) => {
+                    if let Some(delegate) = self.delegates.get_mut(&region_id) {
+                        delegate.update(progress);
+                    } else {
+                        debug!(
+                            "unregistered ReadDelegate, region_id: {}, {:?}",
+                            region_id, progress
+                        );
+                    }
+                }
+                Task::Destroy(region_id) => {
+                    self.delegates.remove(&region_id);
+                }
+            }
+        }
+
+        if let Some(send_time) = sent {
+            self.metrics
+                .requests_wait_duration
+                .observe(duration_to_sec(send_time.elapsed()));
+        }
+    }
+}
+
+const METRICS_FLUSH_INTERVAL: u64 = 15; // 15s
+
+impl<C: Sender<StoreMsg>> RunnableWithTimer<Task, ()> for LocalReader<C> {
+    fn on_timeout(&mut self, timer: &mut Timer<()>, _: ()) {
+        self.metrics.flush();
+        let count = self.delegates.values().fold(0i64, |mut acc, delegate| {
```

Comment:
can we use the delegates count directly, we don't need so much accuracy.

Btw, I worry traversing all delegates here may hurt the performance.

Reply:
All peer(leader/follower) have a read delegates. It collects every 15 seconds, so it can be neglect. Or just remove it? What do you think?

Reply:
em, why delegates contain followers? I think every time a node become the leader, it can register in the delegates, and if the leadership changes, it can remove from the delegates.

Reply:
Because followers can become leaders. It has almost the same lifetime as apply delegate.

Reply:
no, apply is different from this, because follower will also apply raft logs.

Reply:
if follower becomes leader, it can register

Reply:
In current implementation a remote lease sticks to one term, and once we call `expire_remote_lease()` or `expire()` on lease, its remote lease can never renew.

Reply:
A follower can register but it can never use previous remote lease.